### PR TITLE
User authentification

### PR DESCRIPTION
**Pull request by nurbolsultanov**
[Issue #11](https://github.com/JavierAybar/Full_Stack_Capstone_Back_End/issues/11)

In this pull request, I have implemented proper user authentication from the front end to the server. Here are the **key** changes and updates made:

**Redux Integration:** I integrated Redux into the application to manage the state related to user authentication. This includes the creation of an authSlice to handle user authentication actions and a store configuration in src/redux/store.js.

**Authentication Service:** I created an AuthService module (src/services/AuthService.js) to handle user login, registration, and logout operations. This service communicates with the server API endpoints for authentication.

**User Login Page:** Implemented a user login page (src/pages/Login.jsx) where users can enter their email and password to authenticate. Input validation is also included to ensure the entered email is valid.

**User Registration Page:** Created a user registration page (src/pages/Registration.jsx) where users can register by providing their email, password, and password confirmation. Input validation is implemented for user convenience.

**Logout Functionality:** Implemented a logout page (src/pages/Logout.jsx) that logs the user out by communicating with the server to invalidate the authentication token.

**Protected Routes:** Configured protected routes in src/App.js by using the isAuthenticated attribute to restrict access to authenticated users only. This ensures that unauthenticated users are redirected appropriately.

**Redux Toolkit:** Utilized Redux Toolkit for efficient state management and asynchronous actions using createAsyncThunk.

These changes ensure that user authentication is handled securely and efficiently, enhancing the overall functionality of the application.

### CODE REVIEW

**Comment by everinurmind** [comment]
#11

File: src/services/AuthService.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+const API_URL = 'http://localhost:3000/api/v1';
```
Comment (Abdelaziz-Mahdi) [inline src/services/AuthService.js:1]: Hello @everinurmind 
kindly fix the back-end URL as it doesn't include /api/v1
Thanks

File: src/services/AuthService.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  const response = await fetch(`${API_URL}/users/sign_out`, {
+    method: 'DELETE',
+    headers: {
+      'Content-Type': 'application/json',
+    },
+  });
+  if (!response.ok) {
+    throw new Error('Failed to logout');
+  }
+};
```
Comment (Abdelaziz-Mahdi) [inline src/services/AuthService.js]: also, sign-in and sign-out should be something like  ${API_URL}/signin` 
Kindly check back-end routes and correct similar cases

File: src/services/AuthService.js (hunk 1 above)
Comment (Abdelaziz-Mahdi) [inline src/services/AuthService.js]: Hello @everinurmind
kindly fix the back-end URL as it doesn't include /api/v1
Thanks

**Review by Abdelaziz-Mahdi — CHANGES_REQUESTED** [review]
Hello @everinurmind 
### Changes requiered
while you made a great effort small changes are required

- [x] Kindly check back-end routes for example by running the following command `rails routes` and fix all invalid URLs, I have mentioned some of them below.

* Please note that I was able to review only for AuthServices.js

**Comment by JavierAybar** [comment]
Hello @everinurmind 

- Please remember to add the username input on registration page, the user should have `username`, `email`, `password`.
Once you make this change and mege it into `dev `branch, I'll be able to test it well with the rest of the code.
- Please don't forget to resolve branch conflicts.

**Comment by nurbolsultanov** [comment]
@JavierAybar hi, I just did the corrections that you requested. Could you please check it, and merge it if there is no correction needed anymore? Thank you very much!

**Comment by JavierAybar** [comment]
Hello @everinurmind 

Everything looks fine! I need to check it later when the other pr will merge it.
Now I'm going to do the merge.

**Pull request merged**